### ISSUE TITLE
Parameter Binding in Upsert

### ISSUE THREAD

**Issue by chadbean** [ibmdb/node-ibm_db#147]
Parameter binding has been working fine for all of my SELECT, UPDATE, DELETE, and INSERT statements. However, I just tried to use it with an UPSERT (MERGE INTO ...) and I'm getting some odd errors. I should say that if I change this query to use the values directly (without bind params), it works fine. This is on DB2 running on IBM i with a journaled table.

Here is a similar query to what I am using:

```
merge into SCHEMANAME.TABLENAME as tgt
     using (
        values(?, ?)
       ) as src (
        NAME, AGE
      )
      on (tgt.ID) = (src.ID)
      when matched then
        update set 
          tgt.NAME = src.NAME,
          tgt.AGE = src.AGE
      when not matched then
         insert values(src.NAME, src.AGE)
```
Has this module been tested using bind parameters with `MERGE INTO`?

The exact error message I receive is:

```
[IBM][CLI Driver][AS] SQL0584N  The statement failed because of an invalid use of the NULL keyword or the DEFAULT keyword.  SQLSTATE=42608\n
```

**Comment by bimalkjha** [comment]
@chadbean There has been some code change in bind param area to support LOBs. Could you please share an small repro for it? Also, let us know the DDL for SCHEMANAME.TABLENAME, specially the data types of NAME and AGE. Thanks.

**Comment by chadbean** [comment]
Thank you for the reply @bimalkjha. Here is the table schema, with renamed columns (data types are the same).

```
CREATE TABLE "SCHEMA"."TABLENAME" (
		"REFNUM" FOR COLUMN CMREFN NUMERIC(6 , 0) NOT NULL, 
		"STORE" FOR COLUMN CMSTOR NUMERIC(2 , 0) NOT NULL, 
		"NAME" FOR COLUMN CMNAME CHAR(30) CCSID 37 NOT NULL, 
		"PHONE" FOR COLUMN CMPHON1 NUMERIC(10 , 0) NOT NULL, 
		"AVAILALBLE1" COLUMN CMAVAIL1 CHAR(1) CCSID 37 NOT NULL, 
		"DESC1" FOR COLUMN CMDESC1 CHAR(30) CCSID 37 NOT NULL, 
		"DESC2" FOR COLUMN CMDESC2 CHAR(30) CCSID 37 NOT NULL, 
		"AVAILALBLE2" FOR COLUMN CMAVAIL2 CHAR(1) CCSID 37 NOT NULL, 
		"AVAILALBLE3" FOR COLUMN CMAVAIL3 CHAR(1) CCSID 37 NOT NULL, 
		"AVAILALBLE4" FOR COLUMN CMAVAIL4 CHAR(1) CCSID 37 NOT NULL, 
		"DECIMAL1" FOR COLUMN CMDEC1 NUMERIC(4 , 2) NOT NULL, 
		"DECIMAL2" FOR COLUMN CMDEC2 NUMERIC(4 , 2) NOT NULL, 
		"AVAILALBLE5" FOR COLUMN CMAVAIL5 CHAR(1) CCSID 37 NOT NULL, 
		"DECIMAL3" FOR COLUMN CMDEC3 NUMERIC(4 , 2) NOT NULL, 
		"DECIMAL4" FOR COLUMN CMDEC4 NUMERIC(4 , 2) NOT NULL, 
		"AVAILABLE6" FOR COLUMN CMAVAIL6 CHAR(1) CCSID 37 NOT NULL, 
		"DECIMAL5" FOR COLUMN CMDEC5 NUMERIC(4 , 2) NOT NULL, 
		"DECIMAL6" FOR COLUMN CMDEC6 NUMERIC(4 , 2) NOT NULL, 
		"REFID1" FOR COLUMN CMREFID1 INTEGER NOT NULL, 
		"REFID2" FOR COLUMN CMREFID2 BIGINT NOT NULL, 
		"USER" FOR COLUMN CMUSER CHAR(10) CCSID 37 NOT NULL, 
		"CREATEDATE" FOR COLUMN CMCRTDT DATE NOT NULL, 
		"CREATETIME" FOR COLUMN CMCRTTM TIME NOT NULL
	);
```
	
When I insert the VALUES of the query to use ?'s (bind params) and pass in an array of bind parameters it fails with the error I mentioned in the initial post.  If I insert the query without bind parameters and pass in the values with the query, it works fine.

Here is the full function I am using. The `ibm_db` module is being required under the `Db` variable and I am using the `query()` function.

```
const upsert = function (callback) {
  const query = `merge into SCHEMA.TABLENAME as tgt
     using (
      values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, DATE (current timestamp), TIME (current timestamp))
       ) as src (
        REFNUM, STORE, NAME, PHONE, AVAILABLE1, DESC1, DESC2,
        AVAILABLE2, AVAILABLE3, AVAILABLE4, DECIMAL2, DECIMAL1,
        AVAILABLE5, DECIMAL3, DECIMAL4, AVAILABLE6, DECIMAL5, DECIMAL6,
        REFID1, REFID2, CREATEUSER
      )
      on (tgt.REFID1, tgt.REFID2) = (src.REFID1, src.REFID2)
      when matched then
        update set 
          tgt.REFN = src.INVOICE,
          tgt.STORE = src.STORE,
          tgt.NAME = src.NAME,
          tgt.PHONENUMBER = src.PHONENUMBER,
          tgt.AVAILABLE1 = src.AVAILABLE1,
          tgt.DESC1 = src.DESC1,
          tgt.DESC2 = src.DESC2,
          tgt.AVAILABLE2 = src.AVAILABLE2,
          tgt.AVAILABLE3 = src.AVAILABLE3,
          tgt.AVAILABLE4 = src.AVAILABLE4,
          tgt.DECIMAL2 = src.DECIMAL2,
          tgt.DECIMAL1 = src.DECIMAL1,
          tgt.AVAILABLE5 = src.AVAILABLE5,
          tgt.DECIMAL3 = src.DECIMAL3,
          tgt.DECIMAL4 = src.DECIMAL4,
          tgt.AVAILABLE6 = src.AVAILABLE6,
          tgt.DECIMAL5 = src.DECIMAL5,
          tgt.DECIMAL6 = src.DECIMAL6,
          tgt.REFID1 = src.REFID1,
          tgt.REFID2 = src.REFID2,
          tgt.CREATEUSER = src.CREATEUSER
      when not matched then
         insert
            (REFNUM, STORE, NAME, PHONE, AVAILABLE1, DESC1, DESC2,
        	AVAILABLE2, AVAILABLE3, AVAILABLE4, DECIMAL2, DECIMAL1,
        	AVAILABLE5, DECIMAL3, DECIMAL4, AVAILABLE6, DECIMAL5, DECIMAL6,
        	REFID1, REFID2, CREATEUSER)
          values
            (src.REFNUM, src.STORE, src.NAME, src.PHONE, src.AVAILABLE1, src.DESC1, src.DESC2,
       	 	AVAILABLE2, src.AVAILABLE3, src.AVAILABLE4, src.DECIMAL2, src.DECIMAL1,
        	AVAILABLE5, src.DECIMAL3, src.DECIMAL4, src.AVAILABLE6, src.DECIMAL5, src.DECIMAL6,
        	REFID1, src.REFID2, src.CREATEUSER)`;

  const bindingParams = [132358, 12, 'Test Name', '5555551235', 'Y', 'Desc 1',
         'Desc 2', 'Y', 'Y', 'Y', 9.30, 6.00, 'Y', 9.30, 4.00, 'Y', 12.00, 3.00, 12345,
         206, 13850];

  Db.query(query, bindingParams, function (error, data) {
    if (error) return callback(error, null);
    return callback(null, data);
  });
};
```

Please let me know if I can provide anything else. Thanks!

**Comment by chadbean** [comment]
I should also mention that I'm using version 0.0.19 of this package.

**Comment by aaronbartell** [comment]
**Here's a test case that was run on IBM i v7.2:**
```javascript
'use strict';
const Db2 = require('ibm_db');
Db2.debug(true);
const connStr = 'DRIVER={DB2};DATABASE=xxxx;HOSTNAME=99.99.99.99;UID=xxxx;PWD=xxxx;PORT=446;PROTOCOL=TCPIP';
const pool = new Db2.Pool();
const schema = 'S1';
const query = 

// `CREATE SCHEMA ${schema};
`
SET SCHEMA ${schema};
DROP TABLE T1;
CREATE TABLE T1 (COL1 INT, COL2 VARCHAR(24));
MERGE INTO T1 AS T1
  USING ( VALUES(?,?) ) AS SRC(COL1,COL2)
  ON (T1.COL1 = SRC.COL1)
  WHEN MATCHED THEN
    UPDATE SET T1.COL2 = SRC.COL2
  WHEN NOT MATCHED THEN
    INSERT (COL1, COL2) VALUES (SRC.COL1, SRC.COL2)`;
const bindParams = [1,'xxxx'];

pool.open(connStr, function (err, conn) {
    if (err) console.log(err);
    conn.query(query, bindParams, function (err, data) {
        if (err) {
            conn.close(function () {
                console.log(err);
            });
        } else {
            conn.close(function () {
                console.log(data);
            });
        }
    });
});
```
**Here's the output:**
```shell
$ node test.js
node-ibm_db logs enabled.
odbc.js : pool[0] : pool.db.open callback()
Executing
SET SCHEMA S1;
DROP TABLE T1;
CREATE TABLE T1 (COL1 INT, COL2 VARCHAR(24));
MERGE INTO T1 AS T1
  USING ( VALUES(?,?) ) AS SRC(COL1,COL2)
  ON (T1.COL1 = SRC.COL1)
  WHEN MATCHED THEN
    UPDATE SET T1.COL2 = SRC.COL2
  WHEN NOT MATCHED THEN
    INSERT (COL1, COL2) VALUES (SRC.COL1, SRC.COL2)
[]
Pool {
  index: 0,
  availablePool: { 'DRIVER={DB2};DATABASE=xxxx;HOSTNAME=99.99.99.99;UID=xxxx;PWD=xxxx;PORT=446;PROTOCOL=TCPIP': [ [Object] ] },
  usedPool: { 'DRIVER={DB2};DATABASE=xxxx;HOSTNAME=99.99.99.99;UID=xxxx;PWD=xxxx;PORT=446;PROTOCOL=TCPIP': [] },
  odbc: ODBC {} }
[]
Pool {
  index: 0,
  availablePool: { 'DRIVER={DB2};DATABASE=xxxx;HOSTNAME=99.99.99.99;UID=xxxx;PWD=xxxx;PORT=446;PROTOCOL=TCPIP': [ [Object], [Object] ] },
  usedPool: { 'DRIVER={DB2};DATABASE=xxxx;HOSTNAME=99.99.99.99;UID=xxxx;PWD=xxxx;PORT=446;PROTOCOL=TCPIP': [] },
  odbc: ODBC {} }
{ [Error: [IBM][CLI Driver][AS] SQL0584N  The statement failed because of an invalid use of the NULL keyword or the DEFAULT keyword.  SQLSTATE=42608
]
  errors: [],
  error: '[node-odbc] Error in ODBCResult::MoreResultsSync',
  message: '[IBM][CLI Driver][AS] SQL0584N  The statement failed because of an invalid use of the NULL keyword or the DEFAULT keyword.  SQLSTATE=42608\n',
  state: '42608' }
Pool {
  index: 0,
  availablePool: { 'DRIVER={DB2};DATABASE=xxxx;HOSTNAME=99.99.99.99;UID=xxxx;PWD=xxxx;PORT=446;PROTOCOL=TCPIP': [ [Object], [Object], [Object] ] },
  usedPool: { 'DRIVER={DB2};DATABASE=xxxx;HOSTNAME=99.99.99.99;UID=xxxx;PWD=xxxx;PORT=446;PROTOCOL=TCPIP': [] },
  odbc: ODBC {} }
```

**Comment by bimalkjha** [comment]
@aaronbartell  Thanks for sharing the repro steps. I am able to repro the issue using your test program. Investigating it to find the root cause. Thanks.

**Comment by bimalkjha** [comment]
@aaronbartell Issue seems with your query. When I executed same query using ODBC/CLI driver, I can see the same error is coming from server. You can also try below script and see.
```
export IBM_DB_HOME=<path to ibm_db>/installer/clidriver
export PATH=$IBM_DB_HOME/lib:$IBM_DB_HOME/bin:$PATH
export LD_LIBRARY_PATH=$IBM_DB_HOME/lib:$LD_LIBRARY_PATH
db2cli < clitest.t
```
Contents of clitest.t file is as below. Just edit the connection string:
```
opt echo on
opt calldiag on
sqlallocenv 1
sqlallocconnect 1 1
sqldriverconnect 1 0 "DATABASE=LP24UT27;HOSTNAME=LP24UT27.RCH.STGLABS.IBM.COM;PORT=446;PROTOCOL=TCPIP;UID=zurbie;PWD=weR1t3am;" -3 500 SQL_DRIVER_NOPROMPT
sqlallocstmt 1 1
sqlexecdirect 1 "SET SCHEMA S1" -3
sqlexecdirect 1 "DROP TABLE T1" -3
sqlexecdirect 1 "CREATE TABLE T1 (COL1 INT, COL2 VARCHAR(24))" -3
sqlprepare 1 "MERGE INTO T1 AS T1 USING ( VALUES(?,?) ) AS SRC(COL1,COL2) ON (T1.COL1 = SRC.COL1) WHEN MATCHED THEN UPDATE SET T1.COL2 = SRC.COL2 WHEN NOT MATCHED THEN    INSERT (COL1, COL2) VALUES (SRC.COL1, SRC.COL2)" -3
getmem 1 1 SQL_C_LONG
getmem 1 2 SQL_C_CHAR 5
updatemem 1 SQL_C_LONG VALUE 9
updatemem 2 SQL_C_CHAR VALUE 4 xxxx
sqlbindparameter 1 1 SQL_PARAM_INPUT SQL_C_LONG SQL_INTEGER 4 0 1
sqlbindparameter 1 2 SQL_PARAM_INPUT SQL_C_CHAR SQL_VARCHAR 24 0 2
list buffs 1
sqlexecute 1 
sqlexecdirect 1 "select * from T1" -3
fetchall 1 1
sqlfreestmt 1 sql_drop
sqltransact 1 1 sql_rollback
killenv 1
```

Thanks and Regards,
Bimal Jha

**Comment by bimalkjha** [comment]
> sqlexecute 1
SQLExecute: rc = -1 (SQL_ERROR) 
SQLError: rc = 0 (SQL_SUCCESS) 
SQLGetDiagRec: SQLState     : 42608
          fNativeError : -584
          szErrorMsg   : [IBM][CLI Driver][DB2/6000] SQL0584N  The statement failed because of an invalid use of the NULL keyword or the DEFAULT keyword.  SQLSTATE=42608

          cbErrorMsg   : 145
> sqlexecdirect 1 "select * from T1" -3
SQLExecDirect: rc = 0 (SQL_SUCCESS) 
> fetchall 1 1
FetchAll: 0 rows fetched.

**Comment by aaronbartell** [comment]
Thanks, @bimalkjha.

What we're unsure of is *why* the SQL statement is invalid, because when we run the below statement (with literal values, using SQLWorkbenchJ) it runs successfully, and we have run SELECT/INSERT/UPDATE/DELETE with bound parms successfully.  

Is it an accurate statement to say the binding is happening within the DB2 for i engine base on line [937 of odbc.cpp](https://github.com/ibmdb/node-ibm_db/blob/152f93a1ffdd2fdf23d993bc53643de86572920a/src/odbc.cpp#L937)?  

Also, [line 919](https://github.com/ibmdb/node-ibm_db/blob/152f93a1ffdd2fdf23d993bc53643de86572920a/src/odbc.cpp#L919) outputs debug info but that info isn't output when we set `Db.debug(true)`.  How can we have that info be displayed in the console?  Do we need to do our own build of ibm_db as declared in the [build options](https://github.com/ibmdb/node-ibm_db/tree/c37458aefa338eda723a0e877551c7306f6daf5b#build-options)?

```sql
CREATE SCHEMA S1;
SET SCHEMA S1;
CREATE TABLE T1 (COL1 INT, COL2 VARCHAR(24));

MERGE INTO T1 AS T1
  USING ( VALUES(1,'xyz') ) AS SRC(COL1,COL2)
  ON (T1.COL1 = SRC.COL1)
  WHEN MATCHED THEN 
    UPDATE SET T1.COL2 = SRC.COL2
  WHEN NOT MATCHED THEN
    INSERT (COL1, COL2) VALUES (SRC.COL1, SRC.COL2);
``` 

```
Schema S1 created

Execution time: 5.05s
Statement 1 of 4 finished

SET executed successfully

Execution time: 0.14s
Statement 2 of 4 finished

Table T1 created

Execution time: 0.26s
Statement 3 of 4 finished

1 row affected
MERGE executed successfully

Execution time: 0.5s
Statement 4 of 4 finished

Script execution finished
Total script execution time: 5.96s
```

**Comment by bimalkjha** [comment]
#937 of odbc.cpp calls ODBC API SQLBindParameter() to do the parameter binding before calling SQLExecute() to execute the statement. This binding is happening at client side and same data is sent to server when server request for additional data from client after finding parameter marker in SQL. 
Yes, to see the debug info, you need to rebuild the ibm_db using `node-gyp --debug` command. Db.debug(true) is to enable debugging of .js code not the .cpp code. cpp files gets compiled by node-gyp package and generates a new node.js add-on binary. odbc.js file uses this binary to communicate with server.

I used DB2 - CLP to execute your query by giving literal values, so that it became static query and it works. But, when same query we send on server using parameter marker, some how server is not recognising that it needs values from client. We are getting SQL0584N from server. Please contact AS400 server team with the clitest.t file shared in previous post. They should check at server side for root cause of the error. It does not seem an issue with node-ibm_db.  Thanks.

**Comment by aaronbartell** [comment]
**UPDATE**
There were some other fixes put in place for [BLOB insertion](https://github.com/ibmdb/node-ibm_db/issues/120) in June/July and it appears to have addressed this issue also.

**For the archives:** Further testing seemed to show that any time binding parameters were in a second layer SQL statement it would throw the error.  

**I would like to propose closing this issue.**

**Comment by chadbean** [comment]
Thanks, @aaronbartell and @bimaljha. This seems to be fixed by the latest version (1.0), so we'll do some further testing. For now, I'll close.
